context.go

```diff
@@ -2,28 +2,60 @@ package alog
 
 import "context"
 
-type key struct{}
+type stringKey struct{}
 
-var ctxkey = &key{}
+type structuredKey struct{}
+
+var stringCtxKey = &stringKey{}
+var structuredCtxKey = &structuredKey{}
+
+// STag is a structured tag.
+type STag struct {
+	// A unique key for the structure being logged.
+	Key string
+	// The structure you would like logged.  This will be marshalled into an
+	// appropriate form based on the chosen emitter.
+	Val interface{}
+}
 
 // AddTags adds paired strings to the set of tags in the Context.
 //
 // Any unpaired strings are ignored.
 func AddTags(ctx context.Context, pairs ...string) context.Context {
-	old := fromContext(ctx)
+	old := tagsFromContext(ctx)
 	new := make([][2]string, len(old)+(len(pairs)/2))
 	copy(new, old)
 	for o := range new[len(old):] {
 		new[len(old)+o][0] = pairs[o*2]
 		new[len(old)+o][1] = pairs[o*2+1]
 	}
-	return context.WithValue(ctx, ctxkey, new)
+	return context.WithValue(ctx, stringCtxKey, new)
 }
 
-// fromContext wraps the type assertion coming out of a Context.
-func fromContext(ctx context.Context) [][2]string {
-	if t, ok := ctx.Value(ctxkey).([][2]string); ok {
+// tagsFromContext wraps the type assertion coming out of a Context.
+func tagsFromContext(ctx context.Context) [][2]string {
+	if t, ok := ctx.Value(stringCtxKey).([][2]string); ok {
 		return t
 	}
 	return nil
 }
+
+// AddStructuredTags adds tag structures to the Context.
+func AddStructuredTags(ctx context.Context, tags ...STag) context.Context {
+	old := sTagsFromContext(ctx)
+	new := make([]STag, len(old)+len(tags))
+	copy(new, old)
+	for i := range new[len(old):] {
+		new[len(old)+i] = tags[i]
+	}
```

Comment:
Hmm, this is a very awkward way to spell `new = append(old[:len(old):len(old)], tags...)`
(if this were go 1.21+ we could use `new = append(slices.Clip(old), tags...)` or even better `new = slices.Concat(old, tags)` -- https://pkg.go.dev/slices#Clip)

Also, `new` is a keyword, I try to avoid using it as a symbol-name.

Reply:
Hmm -- I guess I didn't really think of it that way.  Your way is definitely more concise, so I went with that.